crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs

```diff
@@ -560,30 +560,54 @@ impl ShardedGossipLocal {
             store::local_agent_arcs(&self.evt_sender, &self.space, &local_agents).await?;
         let arcs: Vec<_> = agent_arcs.iter().map(|(_, arc)| arc.clone()).collect();
         let arcset = local_sync_arcset(arcs.as_slice());
-        let op_hashes = store::all_op_hashes_within_arcset(
-            &self.evt_sender,
-            &self.space,
-            agent_arcs.as_slice(),
-            &arcset,
-            full_time_window(),
-            usize::MAX,
-            true,
-        )
-        .await?
-        .map(|(ops, _window)| ops)
-        .unwrap_or_default();
+        let mut op_hashes = HashMap::new();
+        for (agent, arc) in agent_arcs.clone() {
+            let oh = store::all_op_hashes_within_arcset(
+                &self.evt_sender,
+                &self.space,
+                &[(agent.clone(), arc.clone())],
+                &arcset,
+                full_time_window(),
+                usize::MAX,
+                true,
+            )
+            .await?
+            .map(|(ops, _window)| ops)
+            .unwrap_or_default();
+            op_hashes.insert(agent, (arc, oh));
+        }
 
-        let ops: Vec<_> = store::fetch_ops(
+        let mut needed_op_hashes = HashMap::new();
+        for (agent, (arc, _)) in &op_hashes {
+            for (a, (_, hashes)) in &op_hashes {
+                if a == agent {
+                    continue;
+                }
+                let r: HashSet<_> = hashes
+                    .iter()
+                    .filter(|h| arc.contains(h.get_loc()))
+                    .cloned()
+                    .collect();
+                needed_op_hashes
+                    .entry(agent.clone())
+                    .or_insert_with(HashSet::new)
+                    .extend(r);
+            }
+        }
+
+        let ops_needed: HashSet<_> = needed_op_hashes.values().flatten().cloned().collect();
+        let ops: HashMap<_, _> = store::fetch_ops(
             &self.evt_sender,
             &self.space,
             local_agents.iter(),
-            op_hashes,
+            ops_needed.into_iter().collect(),
+            true,
         )
         .await?
         .into_iter()
         .collect();
+        store::put_ops_direct(&self.evt_sender, &self.space, needed_op_hashes, ops).await?;
```

Comment:
Why wasn't `put_ops` sufficient here? I feel uneasy about doing the arc logic outside of this function and then passing the calculated results directly in. It seems like `put_ops` should be doing this. Why is this needed and do we just need to modify `put_ops` to have logic more like what's above?

Reply:
Because we put_ops combines the results so you have no idea who has which hashes. We need to know which agent has which hashes so we can not send ops to agents that already have them.

Reply:
@freesig then my next question is, is there any reason why we wouldn't always just do it this way? Do we actually want to use `put_ops` in other situations or is it more optimal to just do it this way, removing extraneous puts? I'm questioning whether we can combine these two methods.

Reply:
I'm not sure about this. If we move in the direction I was talking to about today (only having one database) then we always know that the ops are for that one database and don't need to check. It's expensive to pull down all the hashes an agent already has to check if we should send up the ops. This is part of why local sync is expensive. This is also a really good example of how we could simplify gossip if we had a single database.

I think I will leave this with a comment that we might be removing local sync soon and this `put_ops_direct` function soon. Unless you feel I should just remove it from this PR?

Reply:
> If we move in the direction I was talking to about today (only having one database) then we always know that the ops are for that one database and don't need to check

Isn't the point of `put_ops` that it does this extra check? So wouldn't that be an argument in favor of just removing `put_ops`?

Reply:
Because if you have a single database you will never get sent ops you don't already have due to the bloom filter